### ISSUE TITLE
🐛 Timeout parameter is ignored by `run_alternating_checker`

### ISSUE THREAD

**Issue by PabloAndresCQ** [cda-tum/mqt-qcec#477]
### Environment information

- OS: Ubuntu 22.04 (WSL 2.0)
- C++ compiler: g++ 11.4.0
- Python: 3.11.10
- mqt qcec: 2.8.0

### Description

Timeout option seems to be ignored when using the equivalence checker with alternating strategy. See instructions to reproduce below, where I provide two circuits that are equivalent but whose equivalence checking takes very long. Setting a timeout of 5 seconds does not make it abort.

### How to Reproduce

```
from mqt import qcec

result = qcec.verify(
  "circ1.qasm",
  "circ2.qasm",
  run_alternating_checker = True,
  run_construction_checker = False,
  run_simulation_checker = False,
  run_zx_checker = False,
  timeout=5,
)

print(result.equivalence)
```

[circ1.qasm.txt](https://github.com/user-attachments/files/17329443/circ1.qasm.txt)
[circ2.qasm.txt](https://github.com/user-attachments/files/17329445/circ2.qasm.txt)

**Comment by burgholzer** [comment]
Thanks for raising this issue.

Timeouts in QCEC work by checking an atomic flag in between the application of operations. Unfortunately, this means that an operation needs to be fully applied before a timeout can set in. If a certain operation during the equivalence check leads to an explosion in the DD, that operation is highly likely to take a long time to finish.
In that scenario, it will seem like the timeout is being ignored, while in reality it is just waiting for the current operation to finish.
Unfortunately, more fine grained abortion checks would significantly decrease the overall performance.
In addition, there is no clean way to kill a thread without letting it finish its computation. That's something that could be made possible by switching from multi-threading to multi-processing, but the overhead of processes versus threads is huge and that wouldn't really pay off performance wise.

Keeping timeouts is hard...

Any suggestions are welcome. 😌

**Comment by PabloAndresCQ** [comment]
Ah, I see. Your explanation makes sense. I have now tested using very small timeouts, and I can see that indeed they are applied. I also noticed that the process is not responsive to SIGKILL when it gets stuck in a long operation, which I assume comes down to the same issue.

In terms of suggestions:
- Explaining the behaviour of timeout in the docs would be useful (assuming it's not already there and I missed it).
- Perhaps there's a way to extrapolate the timing of the next operation, considering how much the previous ones have taken and the trend of increase in size of the DD?

Cheers!
